File: ios/App/Sources/Features/Chat/Models/FollowUpExtractor.swift
import Foundation

/// Pulls follow-up suggestion lists out of an assistant message so the
/// chat surface can render them as tappable chips below the reply
/// instead of as plain text.
///
/// ## Wire formats supported
///
/// All formats are case-insensitive and may appear anywhere in the
/// message. Multiple markers (even mixed formats) are merged in source
/// order; the first chip is the first suggestion, the last chip is the
/// last suggestion.
///
/// 1. Bracket form (the canonical one we instruct the model to emit):
///
///     ```
///     [SUGGEST: shorter follow-up | a different angle | a code example]
///     ```
///
///    `SUGGEST:` is the keyword; pipe `|` separates labels. Newlines
///    inside a single marker are not allowed (so a stray `]` always
///    closes the marker) but multiple marker lines are.
///
/// 2. Angle-bracket form, used by some fine-tunes that learn XML
///    tool-call conventions from each other:
///
///     ```
///     <suggest>shorter follow-up</suggest>
///     <suggest>a different angle</suggest>
///     ```
///
///    Labels in the angle-bracket form are each label's *own* marker
///    (one suggestion per `<suggest>…</suggest>` block), not pipe
///    separated.
///
/// 3. Numbered list form, used by models that prefer a markdown list
///    over a custom marker:
///
///     ```
///     Next steps:
///     1. Try the smaller API
///     2. Add retry logic
///     3. Switch to streaming
///     ```
///
///    The numbered list detector only fires when (a) at least two
///    consecutive lines match the `^\d+[.)]\s+\S` pattern and (b) the
///    lines appear after a heading line that mentions a follow-up
///    keyword ("next", "follow", "you can", "you might", "try",
///    "consider", "want to", "explore"). Otherwise a benign list in
///    the assistant's prose (e.g. "1. The cause … 2. The fix …") would
///    be misclassified as suggestions.
///
/// Every format strips its marker from the visible content. The chip
/// row is rendered by the view layer (`FollowUpChips`).
public enum FollowUpExtractor {

    /// Suggested max length for a single chip label, so a runaway
    /// model can't fill the chip row with a 2k-character wall of text.
    /// Anything longer is truncated with an ellipsis.
    public static let maxLabelLength = 90

    /// Cap on the total number of chips rendered. Catches pathological
    /// cases (a model emits 200 suggestions) so the chip row doesn't
    /// dominate the screen.
    public static let maxSuggestions = 8

    public struct Result: Equatable {
        /// Visible text with every suggestion marker removed (trimmed).
        public var content: String
        /// Ordered list of suggestion labels.
        public var suggestions: [String]
        /// True if at least one marker was present.
        public var hasSuggestions: Bool { !suggestions.isEmpty }
    }

    // MARK: - Regex patterns

    /// Bracket form: `[SUGGEST: a | b | c]`. Up to 400 chars between
    /// brackets (enough for ~4 reasonable labels) and at most one
    /// bracket pair per line so a stray `]` always closes the marker.
    /// The capture group is the inner labels.
    private static let bracketPattern: NSRegularExpression = {
        try! NSRegularExpression(
            pattern: #"\[SUGGEST:\s*([^\]\n]{1,400})]"#,
            options: [.caseInsensitive]
        )
    }()

    /// Angle-bracket form: `<suggest>label</suggest>`. Each label is
    /// its own block; we don't allow nested tags so user text like
    /// `<suggest>use a <key>thing</key></suggest>` doesn't get
    /// half-eaten.
    private static let anglePattern: NSRegularExpression = {
        try! NSRegularExpression(
            pattern: #"<suggest\b[^>]*>([\s\S]{1,400}?)</suggest>"#,
            options: [.caseInsensitive]
        )
    }()

    /// "Next steps"-style heading. Anchored to a word boundary so
    /// "questionnaire" doesn't trip "next", and matched at the end
    /// of the line so a sentence like "you can also check the docs"
    /// doesn't trigger the heuristic.
    ///
    /// All alternations accept an optional trailing colon / question
    /// mark / period / ellipsis / closing markdown (so a heading
    /// rendered as `**Next steps:**` still matches as a whole line).
    private static let followUpHeadingPattern: NSRegularExpression = {
        try! NSRegularExpression(
            pattern: #"(?im)^[\s*#>\-]*(?:next\s+steps?|follow[\s\-]?ups?|you\s+can(?:\s+also)?|you\s+might(?:\s+want(?:\s+to)?)?|things?\s+to\s+try|to\s+explore|consider(?:\s+trying)?|want\s+to(?:\s+\w+){0,3}?)\s*[:.?!…\-\*]*\s*$"#,
            options: []
        )
    }()

    /// Numbered list item. `1.` or `1)` style with at least one
    /// non-space char of label.
    private static let numberedItemPattern: NSRegularExpression = {
        try! NSRegularExpression(
            pattern: #"^\s*\d{1,2}[.)]\s+(\S.{0,300})"#,
            options: []
        )
    }()

    // MARK: - Public API

    /// Triple-backtick fenced code block. Captures the fence opener
    /// (e.g. "```swift") and the closer ("```") so we can find the
    /// character range of every fenced block and skip it during marker
    /// extraction. Inline single-backtick code spans are not handled
    /// (most models don't emit `[SUGGEST: ...]` inline; the bracket
    /// form is the one that matters here).
    private static let fencedCodeBlockPattern: NSRegularExpression = {
        try! NSRegularExpression(
            pattern: "```[^\n]*\n[\\s\\S]*?```",
            options: []
        )
    }()

    public static func extract(from raw: String) -> Result {
        guard !raw.isEmpty else {
            return Result(content: raw, suggestions: [])
        }

        // Build a "search mask" that hides the contents of every
        // fenced code block. The model can legitimately emit a literal
        // `[SUGGEST: ...]` example inside a code fence, and we don't
        // want to either extract it as a chip or strip the line out
        // of the visible content. We do this by replacing the block
        // body with placeholder whitespace of the same length so
        // character ranges stay aligned with the original string —
        // markers found in `masked` apply to `raw` 1:1, so we run the
        // strip pass on `raw` to keep the code-fence content intact.
        let masked = maskCodeFences(in: raw)

        var collected: [String] = []
        var visible = raw

        // 1. Bracket form first so the marker removal doesn't disturb
        //    angle-bracket or numbered-list boundaries.
        let bracketMatches = bracketPattern.matches(
            in: masked,
            options: [],
            range: NSRange(location: 0, length: (masked as NSString).length)
        )
        if !bracketMatches.isEmpty {
            // First pass: collect labels in source order.
            for match in bracketMatches {
                let body = (masked as NSString).substring(with: match.range(at: 1))
                for raw in body.components(separatedBy: "|") {
                    let label = raw.trimmingCharacters(in: .whitespacesAndNewlines)
                    if !label.isEmpty {
                        collected.append(capped(label))
                    }
                }
            }
            // Second pass: strip markers walking in reverse so each
            // strip doesn't invalidate the ranges that follow.
            for match in bracketMatches.reversed() {
                let ns = visible as NSString
                visible = ns.replacingCharacters(in: match.range, with: "")
            }
        }

        // 2. Angle-bracket form. Same first-pass / reverse-strip
        //    discipline so the body of the last `<suggest>` doesn't
        //    leak across the next marker.
        let angleMatches = anglePattern.matches(
            in: masked,
            options: [],
            range: NSRange(location: 0, length: (masked as NSString).length)
        )
        if !angleMatches.isEmpty {
            for match in angleMatches {
                let body = (masked as NSString).substring(with: match.range(at: 1))
                let trimmed = body.trimmingCharacters(in: .whitespacesAndNewlines)
                if !trimmed.isEmpty {
                    collected.append(capped(trimmed))
                }
            }
            for match in angleMatches.reversed() {
                let ns = visible as NSString
                visible = ns.replacingCharacters(in: match.range, with: "")
            }
        }

        // 3. Numbered list form (only if we haven't already collected
        //    a richer marker — bracket/angle chips are more specific).
        if collected.isEmpty {
            if let (labels, lineRanges, headingRange) = numberedFollowUps(in: masked) {
                collected.append(contentsOf: labels)
                // Strip the heading + the numbered list, in reverse so
                // ranges stay valid.
                let allRanges = (lineRanges + (headingRange.map { [$0] } ?? []))
                    .sorted(by: { $0.location > $1.location })
                for range in allRanges {
                    let ns = visible as NSString
                    visible = ns.replacingCharacters(in: range, with: "")
                }
            }
        }

        // De-dupe while preserving order, then cap. A model that emits
        // both a bracket and a numbered list pointing at the same
        // three options shouldn't render six chips.
        var seen = Set<String>()
        var deduped: [String] = []
        for label in collected {
            let key = label.lowercased()
            if seen.insert(key).inserted {
                deduped.append(label)
                if deduped.count >= maxSuggestions { break }
            }
        }

        return Result(
            content: visible.trimmingCharacters(in: .whitespacesAndNewlines),
            suggestions: deduped
        )
    }

    // MARK: - Helpers

    /// Replace the body of every fenced code block with whitespace
    /// (preserving the fence markers and line breaks so character
    /// offsets stay aligned with `raw`). The bracketed content
    /// disappears from regex scanning but stays in the final
    /// `Result.content` because we work on `masked` for *extraction*
    /// only — the visible-text return value is the un-masked string
    /// minus the markers we found.
    private static func maskCodeFences(in raw: String) -> String {
        let ns = raw as NSString
        let full = NSRange(location: 0, length: ns.length)
        let matches = fencedCodeBlockPattern.matches(in: raw, options: [], range: full)
        guard !matches.isEmpty else { return raw }
        var out = raw as NSString
        for match in matches.reversed() {
            let body = ns.substring(with: match.range)
            // Keep the opener line (e.g. "```swift\n"), blank out the
            // body until just before the closing "```".
            guard let firstNewline = body.firstIndex(of: "\n") else { continue }
            let openerPart = String(body[body.startIndex...firstNewline])
            let bodyStart = body.index(after: firstNewline)
            // Closing fence is the last 3 chars; keep them.
            let innerEnd = body.index(body.endIndex, offsetBy: -3)
            let inner = String(body[bodyStart..<innerEnd])
            let blanked = String(repeating: " ", count: inner.count)
            let replaced = openerPart + blanked + "```"
            out = out.replacingCharacters(in: match.range, with: replaced) as NSString
        }
        return out as String
    }

    /// Truncate a single label to `maxLabelLength` characters with an
    /// ellipsis suffix.
    private static func capped(_ label: String) -> String {
        guard label.count > maxLabelLength else { return label }
        return String(label.prefix(maxLabelLength - 1)) + "…"
    }

    /// Try to detect a numbered-list follow-up section. Returns the
    /// labels (in source order), the per-line ranges, and the heading
    /// range to strip, all matching the (possibly code-fence-masked)
    /// `raw` input.
    private static func numberedFollowUps(in raw: String) -> (labels: [String], ranges: [NSRange], heading: NSRange?)? {
        let ns = raw as NSString
        let full = NSRange(location: 0, length: ns.length)
        // Find the first heading that triggers the follow-up heuristic.
        guard let headingMatch = followUpHeadingPattern.firstMatch(in: raw, options: [], range: full) else {
            return nil
        }
        let headingRange = headingMatch.range
        // Convert the heading range into a line index by counting
        // newlines before it. We also need the line at that index
        // to compute the heading's start offset for the strip pass.
        let prefix = ns.substring(to: headingRange.location + headingRange.length)
        let headingLineEnd = headingRange.location + headingRange.length
        let lines = raw.components(separatedBy: .newlines)
        var charPos = 0
        var headingIndex: Int = -1
        for (i, line) in lines.enumerated() {
            let lineLen = (line as NSString).length
            if charPos + lineLen >= headingLineEnd {
                headingIndex = i
                break
            }
            charPos += lineLen + 1 // +1 for the separator
        }
        guard headingIndex >= 0 else { return nil }

        // From the line after the heading, take any consecutive
        // numbered items. Stop at the first non-numbered, non-blank
        // line so a 3-item follow-up block at the end of a longer
        // reply doesn't get truncated.
        var labels: [String] = []
        var lineRanges: [NSRange] = []
        // Track the character offset of the line we just consumed so
        // we can build a strippable range for each line.
        var offset = 0
        for (i, line) in lines.enumerated() {
            defer { offset += (line as NSString).length + 1 } // +1 for the \n
            if i < headingIndex + 1 { continue }
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.isEmpty { continue }
            let lineNS = line as NSString
            if let m = numberedItemPattern.firstMatch(
                in: line,
                options: [],
                range: NSRange(location: 0, length: lineNS.length)
            ) {
                let label = lineNS.substring(with: m.range(at: 1))
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                if !label.isEmpty {
                    labels.append(capped(label))
                    lineRanges.append(NSRange(location: offset, length: lineNS.length))
                }
                if labels.count >= maxSuggestions { break }
            } else {
                break
            }
        }
        // Need at least 2 numbered items; otherwise we misfired (a
        // "1. The user requested X" inside prose would be a false
        // positive if we accepted a single hit).
        guard labels.count >= 2 else { return nil }
        return (labels, lineRanges, headingRange)
    }
}

File: android/app/src/main/kotlin/app/roamsocket/android/ui/chat/FollowUpExtractor.kt
package app.roamsocket.android.ui.chat

/**
 * Pure-Kotlin port of the iOS `FollowUpExtractor`
 * (`ios/AnyProvCore/Sources/AnyProvCore/Chat/FollowUpExtractor.swift`).
 *
 * Pulls follow-up suggestion lists out of an assistant message so the
 * chat surface can render them as tappable chips below the reply
 * instead of as plain text. Mirrors the iOS contract 1:1 — same wire
 * formats, same chip caps, same code-fence behaviour. The shared
 * parity fixture in `docs/parity/follow-up-cases.json` exercises
 * both sides with the exact same input/output pairs.
 *
 * ## Wire formats supported
 *
 * 1. Bracket form (the canonical one we instruct the model to emit):
 *
 *     ```
 *     [SUGGEST: shorter follow-up | a different angle | a code example]
 *     ```
 *
 * 2. Angle-bracket form, used by some fine-tunes that learn XML
 *    tool-call conventions from each other:
 *
 *     ```
 *     <suggest>shorter follow-up</suggest>
 *     <suggest>a different angle</suggest>
 *     ```
 *
 * 3. Numbered list form, used by models that prefer a markdown list
 *    over a custom marker:
 *
 *     ```
 *     Next steps:
 *     1. Try the smaller API
 *     2. Add retry logic
 *     3. Switch to streaming
 *     ```
 *
 *    The numbered list detector only fires when (a) at least two
 *    consecutive lines match the `^\d+[.)]\s+\S` pattern and (b) the
 *    lines appear after a heading line that mentions a follow-up
 *    keyword ("next", "follow", "you can", "you might", "try",
 *    "consider", "want to", "explore").
 */
object FollowUpExtractor {

    /** Suggested max length for a single chip label. */
    const val MAX_LABEL_LENGTH = 90

    /** Cap on the total number of chips rendered. */
    const val MAX_SUGGESTIONS = 8

    data class Result(
        /** Visible text with every suggestion marker removed (trimmed). */
        val content: String,
        /** Ordered list of suggestion labels. */
        val suggestions: List<String>,
    ) {
        val hasSuggestions: Boolean get() = suggestions.isNotEmpty()
    }

    // Bracket form: [SUGGEST: a | b | c]. Up to 400 chars between
    // brackets and at most one bracket pair per line.
    private val BRACKET_PATTERN: Regex =
        Regex("""\[SUGGEST:\s*([^\]\n]{1,400})]""", RegexOption.IGNORE_CASE)

    // Angle-bracket form: <suggest>label</suggest>. Each label is its
    // own block; we don't allow nested tags.
    private val ANGLE_PATTERN: Regex =
        Regex("""<suggest\b[^>]*>([\s\S]{1,400}?)</suggest>""", RegexOption.IGNORE_CASE)

    // "Next steps"-style heading. Anchored to a word boundary so
    // "questionnaire" doesn't trip "next", and matched at the end of
    // the line so a sentence like "you can also check the docs"
    // doesn't trigger the heuristic.
    private val FOLLOW_UP_HEADING_PATTERN: Regex = Regex(
        """(?im)^[\s*#>\-]*(?:next\s+steps?|follow[\s\-]?ups?|you\s+can(?:\s+also)?|you\s+might(?:\s+want(?:\s+to)?)?|things?\s+to\s+try|to\s+explore|consider(?:\s+trying)?|want\s+to(?:\s+\w+){0,3}?)\s*[:.?!…\-\*]*\s*$"""
    )

    // Numbered list item. 1. or 1) style with at least one non-space
    // char of label.
    private val NUMBERED_ITEM_PATTERN: Regex =
        Regex("""^\s*\d{1,2}[.)]\s+(\S.{0,300})""")

    // Triple-backtick fenced code block. Used to skip the marker
    // inside code fences (so a literal [SUGGEST: …] example in a
    // code block isn't extracted as a chip or stripped from the
    // visible content).
    private val FENCED_CODE_BLOCK_PATTERN: Regex =
        Regex("""```[^\n]*\n[\s\S]*?```""")

    fun extract(raw: String): Result {
        if (raw.isEmpty()) return Result(content = raw, suggestions = emptyList())

        // Mask the body of every fenced code block with whitespace of
        // the same length. Markers found in `masked` map 1:1 to
        // positions in `raw` so we run the strip pass on `raw` to
        // keep the code-fence content intact.
        val masked = maskCodeFences(raw)

        val collected = mutableListOf<String>()
        var visible = raw

        // 1. Bracket form.
        val bracketMatches = BRACKET_PATTERN.findAll(masked).toList()
        if (bracketMatches.isNotEmpty()) {
            // First pass: collect labels in source order.
            for (match in bracketMatches) {
                val body = match.groupValues[1]
                for (rawLabel in body.split('|')) {
                    val label = rawLabel.trim()
                    if (label.isNotEmpty()) {
                        collected.add(capped(label))
                    }
                }
            }
            // Second pass: strip markers walking in reverse so each
            // strip doesn't invalidate the ranges that follow.
            for (match in bracketMatches.reversed()) {
                visible = visible.replaceRange(match.range.first..<match.range.last + 1, "")
            }
        }

        // 2. Angle-bracket form.
        val angleMatches = ANGLE_PATTERN.findAll(masked).toList()
        if (angleMatches.isNotEmpty()) {
            for (match in angleMatches) {
                val body = match.groupValues[1].trim()
                if (body.isNotEmpty()) {
                    collected.add(capped(body))
                }
            }
            for (match in angleMatches.reversed()) {
                visible = visible.replaceRange(match.range.first..<match.range.last + 1, "")
            }
        }

        // 3. Numbered list form (only if we haven't already collected
        //    a richer marker — bracket/angle chips are more specific).
        if (collected.isEmpty()) {
            val numbered = numberedFollowUps(masked)
            if (numbered != null) {
                collected.addAll(numbered.labels)
                // Strip the heading + the numbered list, in reverse so
                // ranges stay valid.
                val allRanges = (numbered.lineRanges + listOfNotNull(numbered.headingRange))
                    .sortedByDescending { it.first }
                for (range in allRanges) {
                    visible = visible.replaceRange(range.first..<range.last + 1, "")
                }
            }
        }

        // De-dupe while preserving order, then cap. A model that
        // emits both a bracket and a numbered list pointing at the
        // same three options shouldn't render six chips.
        val seen = mutableSetOf<String>()
        val deduped = mutableListOf<String>()
        for (label in collected) {
            val key = label.lowercase()
            if (seen.add(key)) {
                deduped.add(label)
                if (deduped.size >= MAX_SUGGESTIONS) break
            }
        }

        return Result(
            content = visible.trim(),
            suggestions = deduped,
        )
    }

    // MARK: - Helpers

    /** Truncate a single label to [MAX_LABEL_LENGTH] with an ellipsis. */
    private fun capped(label: String): String {
        if (label.length <= MAX_LABEL_LENGTH) return label
        return label.substring(0, MAX_LABEL_LENGTH - 1) + "…"
    }

    /**
     * Replace the body of every fenced code block with whitespace of
     * the same length, preserving the fence markers and line breaks
     * so character offsets stay aligned with `raw`. Mirrors the iOS
     * implementation exactly.
     */
    private fun maskCodeFences(raw: String): String {
        val matches = FENCED_CODE_BLOCK_PATTERN.findAll(raw).toList()
        if (matches.isEmpty()) return raw
        // Build by walking the string and replacing fence bodies with
        // whitespace. Indices in `raw` are stable because we keep the
        // fence opener and closer and only blank out the inner body.
        var out = raw
        for (match in matches.reversed()) {
            val body = match.value
            val firstNewline = body.indexOf('\n')
            if (firstNewline < 0) continue
            val openerPart = body.substring(0, firstNewline + 1)
            // Closing fence is the last 3 chars; keep them.
            val innerEnd = body.length - 3
            val inner = body.substring(firstNewline + 1, innerEnd)
            val blanked = " ".repeat(inner.length)
            val replaced = openerPart + blanked + "```"
            out = out.replaceRange(match.range.first..<match.range.last + 1, replaced)
        }
        return out
    }

    /**
     * Try to detect a numbered-list follow-up section in the
     * (possibly code-fence-masked) `raw` input. Returns the labels,
     * the per-line ranges to strip, and the heading range to strip.
     */
    private fun numberedFollowUps(raw: String): NumberedFollowUps? {
        val headingMatch = FOLLOW_UP_HEADING_PATTERN.find(raw) ?: return null
        val headingRange = headingMatch.range
        val lines = raw.split('\n')
        val headingLineEnd = headingRange.last + 1

        // Find the line index of the heading.
        var charPos = 0
        var headingIndex = -1
        for ((i, line) in lines.withIndex()) {
            val lineLen = line.length
            if (charPos + lineLen >= headingLineEnd) {
                headingIndex = i
                break
            }
            charPos += lineLen + 1 // +1 for the \n
        }
        if (headingIndex < 0) return null

        val labels = mutableListOf<String>()
        val lineRanges = mutableListOf<IntRange>()
        var offset = 0
        for ((i, line) in lines.withIndex()) {
            // Defer-style: do the increment at the end of the
            // iteration so `offset` is the start of line i when we
            // record the range.
            val deferOffset = offset
            offset += line.length + 1
            if (i < headingIndex + 1) continue
            if (line.isBlank()) continue
            val match = NUMBERED_ITEM_PATTERN.find(line) ?: break
            val label = match.groupValues[1].trim()
            if (label.isNotEmpty()) {
                labels.add(capped(label))
                lineRanges.add(deferOffset..<deferOffset + line.length)
            }
            if (labels.size >= MAX_SUGGESTIONS) break
        }
        // Need at least 2 numbered items; otherwise we misfired.
        if (labels.size < 2) return null
        return NumberedFollowUps(labels, lineRanges, headingRange)
    }

    private data class NumberedFollowUps(
        val labels: List<String>,
        val lineRanges: List<IntRange>,
        val headingRange: IntRange?,
    )
}
